### PR TITLE
[V3] enforce commands as ours

### PR DESCRIPTION
**Pull request by mikeshardmind**
### Type

- [ ] Bugfix
- [ ] Enhancement
- [ ] New feature

### Description of the changes
Prevents cogs not using our command module from being added (#1880)

### Why in cog load
Rather than do this at the command level, it makes more sense to do at the cog level to prevent the cog from even being partially added.

### CODE REVIEW

**Comment by Jamesp1989SL** [comment]
Forgive me but why is it important to use your system and not the built in ?

**Comment by Tobotimus** [comment]
@useruser3 It's true that this change (especially being a breaking one) deserves some more explanation. I've opened #1972 to shed some light here and perhaps open up some discussion if necessary.

File: redbot/core/bot.py [hunk 1]
```diff
@@ -289,6 +289,16 @@ def unload_extension(self, name):
             if pkg_name.startswith("redbot.cogs"):
                 del sys.modules["redbot.cogs"].__dict__[name]
 
+    def add_cog(self, cog):
+        for attr in dir(cog):
+            if isinstance(discord.ext.commands.Command, attr) and not isinstance(
+                commands.Command, attr
+            ):
```
Comment (Tobotimus) [inline redbot/core/bot.py]: This is just iterating over names of attributes, not the attributes themselves

**Review by Tobotimus — DISMISSED** [review]

**Comment by Redjumpman** [comment]
I think this is a good change. If anything, it helps curb the muscle memory of always using discord.ext. My recommendation is to include a blurb in the docs the reasoning behind why we subclassed `commands.Command`. While on this topic, it might be beneficial to have a section of all Red specific utilities that should be used over standard d.py.

**Comment by mikeshardmind** [comment]
More on the last point by @redjumpman we have a cog cookiecutter repo which will be documented soon to help alleviate any adjustments like this for new cog creators as well

**Review by Redjumpman — CHANGES_REQUESTED** [review]
Running this currently results in a `TypeError`

`TypeError: isinstance() arg 2 must be a type or tuple of types.` 

I tried calling `type()` on attr and the return of `getattr` and was able to get it to move past *that* error. I think because when iterating over dir, not everything is returned as a type. 

However, this does get the runtime error to raise, but I think unexpectedly for the one time we need it.

`RuntimeError: redbot.core.core_commands is not using Red's command module, and cannot be added.`

**Comment by mikeshardmind** [comment]
This should be working all the time now, rather than only in the small toy test case I was using for it before.

File: redbot/core/bot.py [hunk 2]
```diff
@@ -289,6 +289,18 @@ def unload_extension(self, name):
...
                 del sys.modules["redbot.cogs"].__dict__[name]
 
+    def add_cog(self, cog):
+        for attr in dir(cog):
+            _attr = getattr(cog, attr)
+            if isinstance(_attr, discord.ext.commands.Command) and not isinstance(
+                _attr, commands.Command
+            ):
+                raise RuntimeError(
+                    f"{cog.__class__.__name} from {cog.__module__},"
```
Comment (Redjumpman) [inline redbot/core/bot.py]: `cog.__class__.__name` should be `cog.__class__.__name__` otherwise it raises an attribute error.

**Review by Redjumpman — CHANGES_REQUESTED** [review]
After the change I marked is fixed, the message reports as `RuntimeError: RussianRoulette from russianroulette.russianroulette, is not using Red's command module, and cannot be add
ed.` I think this could be written a little more clearly as "The RussianRoulette cog in the russianroulette.russianroulette package must use from redbot.core import commands, instead of from discord.ext import commands." Something along those lines.

**Comment by mikeshardmind** [comment]
@Redjumpman Sorry, I feel like an idiot after that last minute change to give more detailed errors. That's been fixed, and I've added a documentation link to the error, along with a slight tweak to the documentation making it clear that we have additional methods there for red specific things.

**Comment by Redjumpman** [comment]
Awesome. Everything works great. This is now ready for the next release.

**Review by Redjumpman — APPROVED** [review]

**Comment by Tobotimus** [comment]
Thanks for reviewing @Redjumpman, but considering this is a breaking change we haven't announced for yet, and beta 19 being due this Sunday August 12, I'm going to block this until after beta 19.

**Pull request merged**